### PR TITLE
Improve "unrecognized binary" error messages

### PR DESCRIPTION
**Pull request by Timmmm**
Previously it would just print "unrecognized binary" no matter what the underlying error message was. This was partly because it was unable to know the actual file type except by trial and error. This commit uses the magic number instead. It also gives a nicer error for fat Mach-O binaries which are currently unsupported.

### CODE REVIEW

File: internal/binutils/binutils.go [hunk 1]
```diff
@@ -188,13 +186,49 @@ func (bu *Binutils) Open(name string, start, limit, offset uint64) (plugin.ObjFi
...
+	// Read the first 4 bytes of the file.
+
+	f, err := os.Open(name)
+	if err != nil {
+		return nil, fmt.Errorf("error opening %s: %s", name, err)
+	}
+	defer f.Close()
+
+	var header [4]byte
+	_, err = io.ReadFull(f, header[:])
```
Comment (aalexand) [inline internal/binutils/binutils.go]: Merge with next line: `if _, err = ...; err != nil {`

File: internal/binutils/binutils_test.go [hunk 2]
```diff
@@ -361,3 +362,31 @@ func TestLLVMSymbolizer(t *testing.T) {
...
+	}
+
+	if !strings.Contains(err.Error(), "ELF") {
+		t.Errorf("Open malformed ELF error message doesn't contain 'ELF': %v", err)
+	}
+}
+
+func TestOpenMalformedMachO(t *testing.T) {
+	// Test that opening a malformed Mach-O file will report an error containing
+	// the word "ELF".
```
Comment (aalexand) [inline internal/binutils/binutils_test.go]: the word "Mach-O"

File: internal/binutils/binutils_test.go [hunk 3]
```diff
@@ -361,3 +362,31 @@ func TestLLVMSymbolizer(t *testing.T) {
...
+	bu := &Binutils{}
+	_, err := bu.Open(filepath.Join("testdata", "malformed_macho"), 0, 0, 0)
```
Comment (aalexand) [inline internal/binutils/binutils_test.go]: Nit: merge with next line

File: internal/binutils/binutils_test.go [hunk 4]
```diff
@@ -361,3 +362,31 @@ func TestLLVMSymbolizer(t *testing.T) {
 		}
 	}
 }
+
+func TestOpenMalformedELF(t *testing.T) {
+	// Test that opening a malformed ELF file will report an error containing
+	// the word "ELF".
+	bu := &Binutils{}
+	_, err := bu.Open(filepath.Join("testdata", "malformed_elf"), 0, 0, 0)
```
Comment (aalexand) [inline internal/binutils/binutils_test.go]: Nit: merge with next line

File: internal/binutils/binutils_test.go [hunk 5]
```diff
@@ -361,3 +362,31 @@ func TestLLVMSymbolizer(t *testing.T) {
...
+	if err == nil {
+		t.Fatalf("Open malformed ELF succeeded but shouldn't have")
```
Comment (aalexand) [inline internal/binutils/binutils_test.go]: Optional: for this and other error messages below: the common language for test failure messages in Go tests is "Foo: got X, want Y".

File: internal/binutils/binutils_test.go (hunk 4 above)
Comment (Timmmm) [reply internal/binutils/binutils_test.go]: Can't because `err` is used later.

File: internal/binutils/binutils_test.go (hunk 3 above)
Comment (Timmmm) [reply internal/binutils/binutils_test.go]: Can't because `err` is used later.

File: internal/binutils/binutils_test.go (hunk 4 above)
Comment (aalexand) [reply internal/binutils/binutils_test.go]: Ah, sorry missed that. I'll merge once travis is happy (there were some failures, look transient, I restarted the builds).

Thanks again for the PR and bearing with my comments.

Comment (Timmmm) [reply internal/binutils/binutils_test.go]: No problem :-)

**Pull request merged**